app.py

```diff
@@ -702,8 +702,9 @@ def transpile_add_stats(code, level, lang_):
         return result
     except Exception as ex:
         class_name = get_class_name(ex)
+        error_message = hedy_error_to_response(ex)['Error']
         statistics.add(username, lambda id_: DATABASE.add_program_stats(
-            id_, level, number_of_lines, class_name))
+            id_, level, number_of_lines, class_name, error_message))
```

Comment:
why do we need to add the error message here? We have it already in the program, so it can be fetched immediately form the id, in principle?

Reply:
The current dashboard already tracks error messages, so maybe this might even already be stored, but I am not 100% sure how that works.

Reply:
We are indeed tracking the error messages already as exceptions, so the tracking of the program used to look like this:
```
    {
      "id#level": "student1#1",
      "week": "2023-28",
      "id": "student1",
      "level": 1,
      "number_of_lines": 12,
      "successful_runs": 5,
      "IncompleteCommandException": 2,
      "LonelyEchoException": 2
    }
```

And after this PR it looks like this:

```
    {
      "id#level": "student1#1",
      "week": "2023-28",
      "id": "student1",
      "level": 1,
      "number_of_lines": 9,
      "successful_runs": 4,
      "chart_history": [
        1,
        1,
        1,
        1,
        0,
        0,
        0,
        0,
        0,
        0
      ],
      "error_history": [
        "Oops! You forgot a bit of code! On line 10, you need to enter text behind print.",
        "Oops! You forgot a bit of code! On line 10, you need to enter text behind print.",
        "Oops! You forgot a bit of code! On line 10, you need to enter text behind print.",
        "Oops! You forgot a bit of code! On line 10, you need to enter text behind print.",
        "Oops! You forgot a bit of code! On line 10, you need to enter text behind print.",
        "Oops! You forgot a bit of code! On line 10, you need to enter text behind print."
      ],
      "IncompleteCommandException": 6
    }
```

So I guess we are indeed tracking the same thing twice and adding much more weight to the DB? What do you think Miss Felienne?

Reply:
Thanks for clarifying! Which table is that information from? Is it the logs?

I see how we will need to store the information to be able to chart.
It will make the DB a bit bigger but I don't tend to worry about it much. What could be worrisome:

is whether we have to do a lot of fetching this data, that will make the app slower
the fact that we are saving the error message: "Oops! You forgot a bit of code! On line 10" How will that work if people switch languages, or when the teacher has a different language set than a student (which I do)? Can we make that more generic? I am not sure why we even need to store that but I am not very up to date on this PR.

Reply:
That information comes from the table `program-stats`. Also, just did a test setting the language to Spanish to one of the students, and indeed this PR only works for English. That is due to this function:

https://vscode.dev/github/hedyorg/hedy/blob/radboud-new-media-lab/website/statistics.py#L848-L855
```Python
    def misconception_hit(self, error):
        # Currently not in use
        for misconception, keywords in self.misconception_groups.items():
            # Check if the current error is different from the last error;
            # errors that fall in same misconception group are considered same errors
            if error and any(keyword in error.lower() for keyword in keywords):
                return misconception
        return None
```

That checks if the words in the error match a group of missconceptions they created:
```Python
self.misconception_groups = {
            'Not a current level command': ['level'],
            'Incorrect use of command': ['cannot'],
            'Incorrect use of variable': ['variable'],
            'Unwanted spaces': ['Spaces', 'confuse', 'computers'],
            'Forgot commandos': ['forgot'],
            'Empty program': ['empty program'],
            'Typed something that is not allowed': ['entered', 'allowed'],
            'Echo and ask mismatch': ['echo before an ask', 'echo without an ask'],
        }
```

I think maybe we can replace this to use the Exceptions we are already storing, since it will work for every language and also are cheaper to fetch

Reply:
I think this is the only place where they use the exceptions they store in the DB. I do not completely understand the algorithm they used to gather the misconceptions, but I think I can change the way they get a misconception hit

Reply:
> That information comes from the table `program-stats`. 

Got it! That is a nice addition to the data we have here (the last few errors are findable now but it requires a lot of querying) so I am fins with adding this info to the DB.

> Also, just did a test setting the language to Spanish to one of the students, and indeed this PR only works for English. That is due to this function:
> 
> https://vscode.dev/github/hedyorg/hedy/blob/radboud-new-media-lab/website/statistics.py#L848-L855
> 
> ```python
>     def misconception_hit(self, error):
>         # Currently not in use
>         for misconception, keywords in self.misconception_groups.items():
>             # Check if the current error is different from the last error;
>             # errors that fall in same misconception group are considered same errors
>             if error and any(keyword in error.lower() for keyword in keywords):
>                 return misconception
>         return None
> ```
> 
> That checks if the words in the error match a group of missconceptions they created:
> 
> ```python
> self.misconception_groups = {
>             'Not a current level command': ['level'],
>             'Incorrect use of command': ['cannot'],
>             'Incorrect use of variable': ['variable'],
>             'Unwanted spaces': ['Spaces', 'confuse', 'computers'],
>             'Forgot commandos': ['forgot'],
>             'Empty program': ['empty program'],
>             'Typed something that is not allowed': ['entered', 'allowed'],
>             'Echo and ask mismatch': ['echo before an ask', 'echo without an ask'],
>         }
> ```

> I think maybe we can replace this to use the Exceptions we are already storing, since it will work for every language and also are cheaper to fetch

Ah yeah I see, this is almost a 1-1 mapping on our exceptions, so let's match these and make it more language independent and future proof (since we might also change phrasing of messages!)

Reply:
Ok! So I'm going to replace this with the exceptions we are already storing, or something similar, since that is the easiest way to make it language and phrasing independent.